File: algorithms/topological_sort/topological_sort.go
package main

import (
	"fmt"
)

type Edge struct {
	U, V int
}

func KahnsTopologicalSort(n int, edges []Edge) []int {
	adj := make([][]int, n)
	inDegree := make([]int, n)

	for _, e := range edges {
		adj[e.U] = append(adj[e.U], e.V)
		inDegree[e.V]++
	}

	queue := make([]int, 0)
	for i := 0; i < n; i++ {
		if inDegree[i] == 0 {
			queue = append(queue, i)
		}
	}

	order := make([]int, 0, n)
	for len(queue) > 0 {
		u := queue[0]
		queue = queue[1:]
		order = append(order, u)

		for _, v := range adj[u] {
			inDegree[v]--
			if inDegree[v] == 0 {
				queue = append(queue, v)
			}
		}
	}

	if len(order) != n {
		return nil // Cycle
	}
	return order
}

func main() {
	edges := []Edge{
		{5, 2}, {5, 0}, {4, 0}, {4, 1}, {2, 3}, {3, 1},
	}
	n := 6
	order := KahnsTopologicalSort(n, edges)
	if len(order) != n {
		panic("Invalid topological sort result")
	}

	pos := make(map[int]int)
	for i, node := range order {
		pos[node] = i
	}
	for _, e := range edges {
		if pos[e.U] >= pos[e.V] {
			panic(fmt.Sprintf("Topological order violation: %d before %d", e.U, e.V))
		}
	}

	fmt.Printf("[Go TopoSort] Valid ordering verified: %v\n", order)
}

File: algorithms/03_cpp/graphs/topological_sort.cpp
/**
 * Topological Sort (Kahn's Algorithm) in C++.
 */

#include <iostream>
#include <vector>
#include <queue>
#include <cassert>

std::vector<int> kahns_topological_sort(int n, const std::vector<std::pair<int, int>>& edges) {
    std::vector<std::vector<int>> adj(n);
    std::vector<int> in_degree(n, 0);

    for (const auto& edge : edges) {
        adj[edge.first].push_back(edge.second);
        in_degree[edge.second]++;
    }

    std::queue<int> q;
    for (int i = 0; i < n; ++i) {
        if (in_degree[i] == 0) {
            q.push(i);
        }
    }

    std::vector<int> order;
    while (!q.empty()) {
        int u = q.front();
        q.pop();
        order.push_back(u);

        for (int v : adj[u]) {
            if (--in_degree[v] == 0) {
                q.push(v);
            }
        }
    }

    if ((int)order.size() != n) {
        return {}; // Cycle detected
    }
    return order;
}

int main() {
    std::vector<std::pair<int, int>> edges = {
        {5, 2}, {5, 0}, {4, 0}, {4, 1}, {2, 3}, {3, 1}
    };
    int n = 6;
    auto order = kahns_topological_sort(n, edges);
    assert(order.size() == 6);

    std::vector<int> pos(n);
    for (int i = 0; i < n; ++i) {
        pos[order[i]] = i;
    }
    for (const auto& edge : edges) {
        assert(pos[edge.first] < pos[edge.second]);
    }

    std::cout << "[C++ TopoSort] Topological order verified." << std::endl;
    return 0;
}
